components/brave_shields/adblock/rs/src/engine.rs

```diff
@@ -0,0 +1,202 @@
+        request_type: &CxxString,
+        third_party_request: bool,
+        previously_matched_rule: bool,
+        force_check_exceptions: bool,
+    ) -> BlockerResultResult {
+        || -> Result<BlockerResult, InternalError> {
+            let inner_blocker_result = self.engine.check_network_urls_with_hostnames_subset(
+                url.to_str()?,
+                hostname.to_str()?,
+                source_hostname.to_str()?,
+                request_type.to_str()?,
+                Some(third_party_request),
+                previously_matched_rule,
+                force_check_exceptions,
+            );
+            Ok(inner_blocker_result.into())
+        }()
+        .into()
+    }
+
+    pub fn get_csp_directives(
+        &self,
+        url: &CxxString,
+        hostname: &CxxString,
+        source_hostname: &CxxString,
+        request_type: &CxxString,
+        third_party_request: bool,
+    ) -> StringResult {
+        || -> Result<String, InternalError> {
+            Ok(self
+                .engine
+                .get_csp_directives(
+                    url.to_str()?,
+                    hostname.to_str()?,
+                    source_hostname.to_str()?,
+                    request_type.to_str()?,
+                    Some(third_party_request),
+                )
+                .unwrap_or_default())
+        }()
+        .into()
+    }
+
+    pub fn deserialize(&mut self, serialized: &CxxVector<u8>) -> EmptyTupleResult {
+        self.engine.deserialize(serialized.as_slice()).map_err(|e| InternalError::from(e)).into()
+    }
+
+    pub fn add_resource(
+        &mut self,
+        name: &CxxString,
+        content_type: &CxxString,
+        data: &CxxString,
+    ) -> EmptyTupleResult {
+        || -> Result<(), InternalError> {
+            let resource = Resource {
+                name: name.to_str().unwrap().to_string(),
+                aliases: vec![],
+                kind: ResourceType::Mime(MimeType::from(content_type.to_str()?)),
+                content: data.to_string(),
+            };
+            Ok(self.engine.add_resource(resource)?)
+        }()
+        .into()
+    }
+
+    pub fn use_resources(&mut self, resources_json: &CxxString) -> EmptyTupleResult {
+        || -> Result<(), InternalError> {
+            let resources: Vec<Resource> = serde_json::from_str(resources_json.to_str()?)?;
+            Ok(self.engine.use_resources(&resources))
+        }()
+        .into()
+    }
+
+    pub fn url_cosmetic_resources(&self, url: &CxxString) -> StringResult {
+        || -> Result<String, InternalError> {
+            let resources = self.engine.url_cosmetic_resources(url.to_str()?);
+            Ok(serde_json::to_string(&resources)?)
+        }()
+        .into()
+    }
+
+    pub fn hidden_class_id_selectors(
+        &self,
+        classes: &CxxVector<CxxString>,
+        ids: &CxxVector<CxxString>,
+        exceptions: &CxxVector<CxxString>,
+    ) -> VecStringResult {
```

Comment:
`classes` contains HTML/CSS classes queried from elements on a loaded page.
`ids` likewise contains HTML/CSS ids queried from elements on a loaded page.
`exceptions` is always provided directly from `url_cosmetic_resources`, unmodified.

`exceptions` is definitely safe, and... I _think_ HTML classes and ids should always be valid UTF-8? But I'm fine to leave the result here if we're not 100% sure about that. Especially considering how clean `convert_cxx_string_vector_to_string_collection` is :smile: